### DISCUSSION TITLE
xdrv_17_rcswitch.ino pulselen overwritten with rfsend command

### DISCUSSION THREAD

**Discussion by PepeTheFroggie** [arendst/Tasmota#10864] [Ideas]
rcswitch:
Pulselen is a part of the protocol and should not be overwritten if not specified in the rfsend command. 
Overwriting it will result in the wrong (protocol1 default) pulselen for everyone not aware of the issue.

The lines 198 and 199:
`if (!pulse) { pulse = 350; }      // Default pulse length for protocol 1`
`    mySwitch.setPulseLength(pulse);`
should be replaced with:
`if (pulse != 350) { mySwitch.setPulseLength(pulse); }`

Pulse is initialized to 350 and guess what effect the `if (!pulse) ` then has? Yes, none. If you want to use (!pulse) you cant init the value.
Same with protocol, repeat, bits, if you init them you cant ask `if (!repeat) { repeat = 10; }`

The whole section should look like:
![Clipboard01](https://user-images.githubusercontent.com/37778140/107137165-f589c100-6909-11eb-9e24-f0dc718b6ac4.png)

**Comment by barbudor** [comment]
As per the known lyrics: "I don't subscribe to your point of view"

Although pulse is initialized to 350 at the beginning of the function, then it is read from the command line following one of 2 format : JSON or comma-separated.
The check you are pointing to to test against an invalid zero value before passing pulse to the library.

The simplified whole sequence is:
```
    int pulse = 350;
    if (root) {    // RFsend {"data":0x501014,"bits":24,"protocol":1,"repeat":10,"pulse":350}
      pulse = root.getInt(PSTR(D_JSON_RF_PULSE), pulse);
    } else {       //  RFsend data, bits, protocol, repeat, pulse
      pulse = atoi(str);
    }
    if (!pulse) { pulse = 350; }      // Default pulse length for protocol 1
    mySwitch.setPulseLength(pulse);
```
The 1st case get pulse form a JSON. Providing the current pulse value (350) insure that pulse always gets a good value even if not present in the JSON.
In the 2nd case, `pulse = atoi(str);` reads from the command line. Behavior of `atoi()` is to return 0 if the string is not a valid number.
Then, the one before last 2 lines properly identify 0 as an wrong value and set back pulse to the default 350.
And finally the pulse value is passed to the library.

With your code, pulse would be set to the library if it is different than 350.
Especially it would allow setting an illegal value of 0.
This sounds totally unsafe.

May be you can elaborate on the use-case where the current code makes a problem ?

**Reply by PepeTheFroggie** [reply]
Let me express myself a little better.

The code mySwitch.setProtocol(protocol); does already set the correct pulse for the protocol.
Now make sure you do not overwrite that value if the user does not explicitly wants this.

Today the command 
RFsend {"Data":"0x000014","Bits":24,"Protocol":4,"Repeat":5} 
DOES overwrite pulse with the wrong value.

**Reply by barbudor** [reply]
OK, with the explanations it makes sense now

So it looks like the proper logical sequence should be:
```
    int pulse = 0;  // 0 means let the library use the default depending on the protocol
    if (root) {    // RFsend {"data":0x501014,"bits":24,"protocol":1,"repeat":10,"pulse":350}
      pulse = root.getInt(PSTR(D_JSON_RF_PULSE), pulse);  // leave 0 if not specified
    } else {       //  RFsend data, bits, protocol, repeat, pulse
      pulse = atoi(str);  // would be 0 if string can't convert to number
    }
    if (pulse)  // non-zero means value specified on the command, so use it (otherwise let the lib use its own default)
        mySwitch.setPulseLength(pulse);
```

Does it sounds better ?

**Reply by PepeTheFroggie** [reply]
Not tested but from eyeballing 100% correct !

**Reply by barbudor** [reply]
Can you make a PR ?
If not, can you at least test and I could make the PR

**Reply by PepeTheFroggie** [reply]
The pulse thing was backwards and "bits" and "repeat" are not part of protocol so always set them.

Could not figure out how to paste code here. I attach a text file.
[code.txt](https://github.com/arendst/Tasmota/files/5938880/code.txt)

**Reply by barbudor** [reply]
Correct, my mistake
So the attached is working ?
I'll make the PR on your name

**Reply by PepeTheFroggie** [reply]
TY.

Sorry, im quite old and i struggle with PR and git in general.

**Reply by barbudor** [reply]
https://github.com/arendst/Tasmota/pull/10866

**Reply by PepeTheFroggie** [reply]
Still got the (!pulse) wrong. Must be (pulse)
not:  if (!pulse) { mySwitch.setPulseLength(pulse); } //(pulse == 0)
but:  if (pulse) { mySwitch.setPulseLength(pulse); } //(pulse != 0)

**Reply by barbudor** [reply]
Correct
New PR : https://github.com/arendst/Tasmota/pull/10876
